app/modules/utils/rate_limiter.py

```
@@ -0,0 +1,44 @@
+
+        async with self._rate_limit_lock:
+            now = datetime.now()
+            # Remove timestamps older than 1 minute
+            while self._request_timestamps and self._request_timestamps[0] < now - timedelta(minutes=1):
+                self._request_timestamps.popleft()
+
+            # Check if we're at the rate limit
+            if len(self._request_timestamps) >= self.MAX_REQUESTS_PER_MINUTE:
+                wait_time = (self._request_timestamps[0] + timedelta(minutes=1) - now).total_seconds()
+                if wait_time > 0:
+                    logging.warning(f"Rate limit reached for {self.name}. Waiting {wait_time:.2f} seconds")
```

Ruff (0.8.2)

Undefined name `logging`

(F821)